### PR TITLE
Updated compute, a2a and dispatch tests to work with 1x8 mesh shape

### PR DESCRIPTION
**Pull request by smeydanshahiTT**
### Ticket
N/A

### Problem description
MoE tests were previously hardcoded to only support 1x16 mesh configs for Galaxy. This stopped us from running this tests on systems with different mesh configs. We wanted to be able to use 1x8, so we can test more frequently. 

### What's changed
- Extended MoE tests (compute, combine, dispatch) to support both 1x8 and 1x16 mesh configs
- Added a new mesh descriptor file, defining a 1x8 torus topology for a single galaxy
- Parameterized experts per device across all tests
- total experts is now calculated dynamically using the mesh shape 
- Updated the matmul dm1 kernel to maintain the buffer size of `512` for both 1x16 and 1x8 configs
- Added assertion to validate buffer size matches tensor dimensions

Tests can now run on both 1x8 and 1x16 mesh configs, and the code has become more readable and maintainable.

### Checklist

- [ ] [Sanity Tests](https://github.com/tenstorrent/tt-metal/actions/runs/23017540965)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/23345268263)
- [ ] [Galaxy e2e](https://github.com/tenstorrent/tt-metal/actions/runs/23017404607)
- [ ] [DeepSeek V3 dual](https://github.com/tenstorrent/tt-metal/actions/runs/22922010194)
- [ ] [DeepSeek V3 quad](https://github.com/tenstorrent/tt-metal/actions/runs/22922029958)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/experimental/ccl/moe_compute/device/moe_compute_program_factory.cpp [hunk 1]
```diff
@@ -932,7 +932,7 @@ MoEComputeMeshWorkloadFactory::create_at(
         {"tile_height", tile_height},
         {"tile_width", tile_width},
         {"tile_width_size_bytes", tile_width * tt::datum_size(tilize_output_dataformat)},
-        {"num_tokens_total", tokens},
+        {"buffer_size_total_tokens", 512},  // Hardware buffer is always sized for 512 tokens
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/experimental/ccl/moe_compute/device/moe_compute_program_factory.cpp]: hm any reason why just tokens didn't work (with an assert for anything over 512)

Comment (smeydanshahiTT) [reply ttnn/cpp/ttnn/operations/experimental/ccl/moe_compute/device/moe_compute_program_factory.cpp]: Yea so I believe the output buffer is allocated with that size regardless of whether we're running on a 1x8 or 1x16. So for context, this `num_tokens_total` value was previously set in the test to equal `total_tokens`, which was calculated as `tokens_per_device * num_devices`. In the `test_moe_compute_6U.py` file, theres a function that tries to reshape the output, and the buffer matches the expected shape as the shape is calculated as `512 // 4 = 128`. However, when the number of devices is 8, this num tokens value became 256, thus during that reshape there was no match to the output shape.

@amorrisonTT can also comment on this.

Comment (amorrisonTT) [reply ttnn/cpp/ttnn/operations/experimental/ccl/moe_compute/device/moe_compute_program_factory.cpp]: Compute output and Combine are reusing the same buffer that was allocated for the output of tilize/input of compute which is always 32x7168 (tile size x hidden). Combine needed 512x1792 (num tokens, x hidden/4), but for smaller batch sizes this buffer still needs to be 512x1792, regardless of num tokens, to match the tilize buffer volume.

Comment (amorrisonTT) [reply ttnn/cpp/ttnn/operations/experimental/ccl/moe_compute/device/moe_compute_program_factory.cpp]: But also, make this a named `constexpr` rather than int literal

File: tests/nightly/tg/ccl/moe/test_moe_compute_6U.py [hunk 2]
```diff
@@ -270,20 +272,22 @@ def prepare_output_tensor_from_combine_writer(
     combine_output_shards = [all_output_shards[c.x, c.y] for c in output_shard_cores]
     output_shard_tensor = torch.stack(combine_output_shards)
 
+    buffer_size_total_tokens = 512
+
```
Comment (amorrisonTT) [inline tests/nightly/tg/ccl/moe/test_moe_compute_6U.py:262]: You can codify the assumption here with something like,
`assert output_shard_tensor.shape[1] ==  buffer_size_total_tokens` although I might be wrong about the dims off the top of my head.

File: tests/nightly/tg/ccl/moe/test_selective_combine_6U.py [hunk 3]
```diff
@@ -675,6 +712,7 @@ def test_decode(
     num_test_iters,
     num_inner_iters,
 ):
+    experts = 2 * mesh_shape[1]  # 2 * 8 for 1x8 mesh, 2 * 16 for 1x16 mesh
```
Comment (grechsteinerTT) [inline tests/nightly/tg/ccl/moe/test_selective_combine_6U.py]: Probably best to parameterize the 2 as `experts_per_device`, and use the cluster axis in the total experts calculations. Can do a similar thing in the other test files as well.

i.e. `experts = experts_per_device * mesh_shape[cluster_axis]`

File: tests/nightly/tg/ccl/moe/test_moe_compute_6U.py (hunk 2 above)
Comment (smeydanshahiTT) [reply tests/nightly/tg/ccl/moe/test_moe_compute_6U.py:262]: none of the dims of `output_shard_tensor` are = to 512 though? in both the 1x16 and 1x8 case `output_shard_tensor.shape = [16, 2, 32, 7168]`

Comment (amorrisonTT) [reply tests/nightly/tg/ccl/moe/test_moe_compute_6U.py:262]: True it's a little more involved, something like `output_shard_tensor.shape[-1] * output_shard_tensor.shape[-2] // hidden_size // output_shard_width_dim` or something. I just think that validating the assumption with a commented assert statement is more illuminating to the future user than a the `view` failing for some unclear reason

**Review by sjameelTT — APPROVED** [review]

**Review by amorrisonTT — APPROVED** [review]

**Review by ntarafdar — CHANGES_REQUESTED** [review]
please update the PR description and issue. 

Also maybe outside the scope of this but can we see if we can get a 4x8 mesh, that way we can also test multiple rows.

@amorrisonTT can probably help there

File: tests/tt_metal/tt_fabric/custom_mesh_descriptors/single_galaxy_1x8_torus_graph_descriptor.textproto [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# --- Meshes ---------------------------------------------------------------
+
+mesh_descriptors {
+  name: "M0"
+  arch: WORMHOLE_B0
+  device_topology { dims: [ 1, 8 ] dim_types: [ RING, RING ] }
```
Comment (SeanNijjar) [inline tests/tt_metal/tt_fabric/custom_mesh_descriptors/single_galaxy_1x8_torus_graph_descriptor.textproto:6]: dim0 as ring doesn't make sense in a 1x8 on a galaxy. @Riddy21 can you comment if this will be implicitly converted to line on dim0?

**Review by ntarafdar — APPROVED** [review]

File: tests/tt_metal/tt_fabric/custom_mesh_descriptors/single_galaxy_1x8_torus_graph_descriptor.textproto (hunk 4 above)
Comment (Riddy21) [reply tests/tt_metal/tt_fabric/custom_mesh_descriptors/single_galaxy_1x8_torus_graph_descriptor.textproto:6]: I'm pretty sure I ignore self connections when doing topology mapping, so these would not be represented in the routing tables. You can try running with this, if it passes topology mapper then it should work, but if you come up with some constraint error or topology mapper error, just change device_topology for dim0 to LINE

Comment (smeydanshahiTT) [reply tests/tt_metal/tt_fabric/custom_mesh_descriptors/single_galaxy_1x8_torus_graph_descriptor.textproto:6]: @SeanNijjar I haven't encountered any such errors, so based on Ridvan's comment I think the descriptor should be fine

**Review by SeanNijjar — APPROVED** [review]
